cypress-io/cypress
# feat: experimental retries

- Closes N/A

## Description

This feature branch his based of the `feature/test-burn-in` branch with the following commits reverted:
* c42844307946f438a7369b381a702f121ff17657 which introduces the experimentalBurnIn config option ([PR](https://github.com/cypress-io/cypress/pull/27412)). `experimentalBurnIn` is not ready for release yet and we have decided to omit the config options from the release of `experimentalRetries` 
* ae3df1a505ed16d29e524a941078ea28b0a94b3c ([PR](https://github.com/cypress-io/cypress/pull/27768)) which removes the cloud `burnInAction` as it isn't needed for experimentalRetries release.
* 83842a6f82 Removes the update the the `experimentalBurnIn` snapshot which was updated with the v13 release

ALL COMMITS ON THIS BRANCH HAVE BEEN PREVIOUSLY REVIEWED (with the exception of f010aef01e80965ecbc1d1c01f9e506b1c398d53)!

### Goals / Risk

The goal of this feature branch is to release `experimentalRetries` ahead of `experimentalBurnIn`, which will allow users to try the new configuration options for themselves and report any bugs if found.

The secondary goal of releasing `experimentalRetries` ahead of burn in is to get users to experiment with the experimental feature and identify bugs that might occur with `experimentalRetries`. However, the primary goal is to detect any issues that might have occurred adapting the existing GA `retries` to the new retries engine implemented under the hood of the Cypress App. 

It's important to understand that in order to support `experimentalRetries`, we needed to patch `mocha` in order to rerun passed test attempts. To accomplish this, we needed to change how retries fundamentally work, which means signficant changes to the retries mechanism. In short, this means that the logic within the new `calculateTestStatus` function backports the retries functionality into the new engine. An easy way to think about this would be:
* A GA config of `retries: 2` would map to a strategy of `detect-flake-and-pass-on-threshold` with a `maxRetries` of `2` and a `passesRequired` of `1`

In addition, the new field, called `_cypressTestStatusInfo`, which is used by both the cypress reporter in console and in the browser, is still used in GA implementation to interpret the test outerStatus, and will fallback to the status of the last attempt.

It's important to note that there isn't currently anything signaling that the implementation changes are an issue, but we want to ship this out to a wider audience to reduce the scope of changes and have the impact be deterministic. If issues occur with the release of this feature, we can determine those issues with as little noise as possible, and if needed, rollback easily and assess what went wrong. 

For more information on this, please see the following PRs that go into the implementation details more specifically:
* https://github.com/cypress-io/cypress/pull/27412 implements the experimental retries config options
* https://github.com/cypress-io/cypress/pull/27826 implementation of experimental retries


### The Feature


Experimental retries will allow users to leverage what we are calling a retry strategy. These strategies include
* `detect-flake-and-pass-on-threshold`
* `detect-flake-but-always-fail`

`detect-flake-and-pass-on-threshold` is similar to how retries work today. Users will be able to specify a retry limit and specify a required number of passes needed in order for that to be marked as passed. Today, that limit is `1`, but users can set the limit to whatever they'd like. This can be useful if a test is retrying and is known to be flaky, but the test engineers wishes to make sure that the test can at least pass a certain threshold in order to be marked as passed, which might help in confidence that the test is just flaky and the failure is not legitimate. It's important to note that if the `passesRequired` can not be acheived by the remaining retries of the test, the test will stop retrying and the test will be marked as a failure.

The other, `detect-flake-but-always-fail`, will always mark a test that enters retries as failed, regardless of how many retried attempts pass. There is an option here, called `stopIfAnyPassed`, that will stop the test retries if any attempt passes. Without this option, a user might want to see what the probability that their test might pass (similar to `detect-flake-and-pass-on-threshold`, but mark the test as failed as the test engineer might not want to mark any detected flaky test as a passed test.

##### with `experimentalStrategy` and `experimentalOptions` enabled
![](https://github.com/cypress-io/cypress/assets/3980464/5e23fc4b-bb7c-4b1f-80a0-6111419e0415)

##### with only `experimentalStrategy` enabled
![](https://github.com/cypress-io/cypress/assets/3980464/7a36f150-547d-48fd-bbf4-85559cf54545)

##### with neither enabled
![](https://github.com/cypress-io/cypress/assets/3980464/dacd8990-d152-40a1-8a3a-4661556a358a)

For reference, here are some configuration examples as to hopw the experimental options compare to what exists today, and how their outputs differ. 

#### current implementation (GA)

```js
retries: 2
```

![](https://github.com/cypress-io/cypress/assets/3980464/e2e030a1-48ab-4a76-8fe6-1ce5de3548b5)

#### Experimental options
##### Detect Flake and Pass on Threshold

```js
experimentalStrategy: 'detect-flake-and-pass-on-threshold'
experimentalOptions: {
  maxRetries: 9,
  passesRequired: 5
}
```
![](https://github.com/cypress-io/cypress/assets/3980464/aaeadf33-06d7-4f8a-9c51-2ab4be0adad1)

##### Detect Flake but Always Fail

```js
experimentalStrategy: 'detect-flake-but-always-fail'
experimentalOptions: {
  maxRetries: 5,
  stopIfAnyPassed: false
}
```
![](https://github.com/cypress-io/cypress/assets/3980464/3873c8ea-85d1-4150-9787-06a0485f5736)

##### Detect Flake but Always Fail (`stopIfAnyPassed=true`)

```js
experimentalStrategy: 'detect-flake-but-always-fail'
experimentalOptions: {
  maxRetries: 10,
  stopIfAnyPassed: true
}
```
![](https://github.com/cypress-io/cypress/assets/3980464/0c8fb5d2-aa8c-4890-ab8c-b122b41dd7a7)


### Steps to test
For how this experiment is tested, please see the description of https://github.com/cypress-io/cypress/pull/27826, which uses several testing strategies and mechanisms to guarantee functionality

### How has the user experience changed?
Users will now be able to provide `experimentalStrategy` and `experimentalOptions` in their global config and watch the new form of retries take hold in their test

### PR Tasks
